src/Java.Interop.Tools.JavaTypeSystem/Adapters/JavaXmlApiImporter.cs

```diff
@@ -0,0 +1,370 @@
+				value: element.Attribute ("value")?.Value,
+				parent: type,
+				isFinal: element.XGetAttributeAsBool ("final"),
+				deprecated: element.XGetAttribute ("deprecated"),
+				jniSignature: element.XGetAttribute ("jni-signature"),
+				isTransient: element.XGetAttributeAsBool ("transient"),
+				isVolatile: element.XGetAttributeAsBool ("volatile"),
+				isNotNull: element.XGetAttributeAsBool ("not-null")
+			);
+
+			if (element.XGetAttribute ("merge.SourceFile") is string source && source.HasValue ())
+				field.PropertyBag.Add ("merge.SourceFile", source);
+			if (element.XGetAttribute ("deprecated-since") is string dep && dep.HasValue ())
+				field.PropertyBag.Add ("deprecated-since", dep);
+
+			return field;
+		}
+
+		public static JavaImplementsModel ParseImplements (XElement element)
+		{
+			var model = new JavaImplementsModel (
+				name: element.XGetAttribute ("name"),
+				nameGeneric: element.XGetAttribute ("name-generic-aware"),
+				jniType: element.XGetAttribute ("jni-type")
+			);
+
+			if (element.XGetAttribute ("merge.SourceFile") is string source && source.HasValue ())
+				model.PropertyBag.Add ("merge.SourceFile", source);
+
+			return model;
+		}
+
+		public static JavaExceptionModel ParseException (XElement element)
+		{
+			return new JavaExceptionModel (
+				name: element.XGetAttribute ("name"),
+				type: element.XGetAttribute ("type-generic-aware")
+			);
+		}
+
+		public static JavaParameterModel ParseParameter (JavaMethodModel method, XElement element)
+		{
+			var parameter = new JavaParameterModel (
+				parent: method,
+				javaName: element.XGetAttribute ("name"),
+				javaType: element.XGetAttribute ("type"),
+				jniType: element.XGetAttribute ("jni-type"),
+				isNotNull: element.XGetAttributeAsBool ("not-null")
+			);
+
+			return parameter;
+		}
+
+		public static void ParseTypeParameters (JavaTypeParameters parameters, XElement element)
+		{
+			foreach (var elem in element.Elements ()) {
+				if (elem.Name.LocalName == "typeParameter")
+					ParseTypeParameter (parameters, elem);
+			}
+
+			if (element.XGetAttribute ("merge.SourceFile") is string source && source.HasValue ())
+				parameters.PropertyBag.Add ("merge.SourceFile", source);
+		}
+
+		public static void ParseTypeParameter (JavaTypeParameters parameters, XElement element)
+		{
+			var parameter = new JavaTypeParameter (element.XGetAttribute ("name"), parameters) {
+				ExtendedJniClassBound = element.XGetAttribute ("jni-classBound"),
+				ExtendedClassBound = element.XGetAttribute ("classBound"),
+				ExtendedInterfaceBounds = element.XGetAttribute ("interfaceBounds"),
+				ExtendedJniInterfaceBounds = element.XGetAttribute ("jni-interfaceBounds")
+			};
+
+			parameters.Add (parameter);
+
+			if (element.Element ("genericConstraints") is XElement gc) {
+				parameter.GenericConstraints.AddRange (ParseGenericConstraints (gc));
+				return;
+			}
+
+			// Now we have to deal with the format difference...
```

Comment:
*Do* we need to deal with format differences?  Why?

Though the fact that you had to do something is unexpected means that it *is* needed, but I still don't understand *why*.

My expectation is that there are only two sources of `Java.Interop.Tools.JavaTypeSystem` model data:

 1. Java bytecode, via `.jar` files, `.aar` files (which contain `.jar` files), etc., and
 2. Existing binding assemblies.

This particular codepath only deals with (1), for which you should only need to worry about one `Xamarin.Android.Tools.Bytecode.dll` output format: the one distributed alongside `Java.Interop.Tools.JavaTypeSystem.dll`, i.e. the same one in this tree.

I don't currently understand *how* `Java.Interop.Tools.JavaTypeSystem.dll` could be consuming `api.xml` output generated by a *different* version of `class-parse`/`Xamarin.Android.Tools.Bytecode.dll` (short of "mucking about with" an installation), and thus don't understand why this code needs to support consuming multiple `class-parse` output formats.

Reply:
🤷 

This is c/p from what we currently have: https://github.com/xamarin/java.interop/blob/a3de91efdaec04b3927b755a07018e8d26cfdc71/src/Xamarin.Android.Tools.ApiXmlAdjuster/JavaApiXmlLoaderExtensions.cs#L360.

Leaving it seems safer than trying to figure out if anything in the wild relies on whatever this was added for.